api/client/client.go

```diff
@@ -0,0 +1,230 @@
+	if err != nil {
+		return err
+	}
+
+	for i, entity := range resp.GetEntities() {
+		err := json.Unmarshal([]byte(entity), items[i])
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// ModelSave saves existing instances
+func (c *Client) ModelSave(storeID, modelName string, items ...interface{}) error {
+	values, err := marshalItems(items)
+	if err != nil {
+		return err
+	}
+
+	req := &pb.ModelSaveRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		Values:    values,
+	}
+	_, err = c.client.ModelSave(c.ctx, req)
+	return err
+}
+
+// ModelDelete deletes data
+func (c *Client) ModelDelete(storeID, modelName string, entityIDs ...string) error {
+	req := &pb.ModelDeleteRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityIDs: entityIDs,
+	}
+	_, err := c.client.ModelDelete(c.ctx, req)
+	return err
+}
+
+// ModelHas checks if the specified entities exist
+func (c *Client) ModelHas(storeID, modelName string, entityIDs ...string) (bool, error) {
+	req := &pb.ModelHasRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityIDs: entityIDs,
+	}
+	resp, err := c.client.ModelHas(c.ctx, req)
+	if err != nil {
+		return false, err
+	}
+	return resp.GetExists(), nil
+}
+
+// ModelFind finds records by query
+func (c *Client) ModelFind(storeID, modelName string, query es.JSONQuery) ([][]byte, error) {
+	queryBytes, err := json.Marshal(query)
+	if err != nil {
+		return [][]byte{}, err
+	}
+	req := &pb.ModelFindRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		QueryJSON: queryBytes,
+	}
+	resp, err := c.client.ModelFind(c.ctx, req)
+	if err != nil {
+		return [][]byte{}, err
+	}
+	return resp.GetEntities(), nil
+}
+
+// ModelFindByID finds a record by id
+func (c *Client) ModelFindByID(storeID, modelName, entityID string, entity interface{}) error {
+	req := &pb.ModelFindByIDRequest{
+		StoreID:   storeID,
+		ModelName: modelName,
+		EntityID:  entityID,
+	}
+	resp, err := c.client.ModelFindByID(c.ctx, req)
+	if err != nil {
+		return err
+	}
+	err = json.Unmarshal([]byte(resp.GetEntity()), entity)
+	return err
+}
+
+// ReadTransaction returns a read transaction that can be started and used and ended
+func (c *Client) ReadTransaction(storeID, modelName string) (*ReadTransaction, error) {
+	client, err := c.client.ReadTransaction(c.ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &ReadTransaction{client: client, storeID: storeID, modelName: modelName}, nil
+}
+
+// WriteTransaction returns a read transaction that can be started and used and ended
+func (c *Client) WriteTransaction(storeID, modelName string) (*WriteTransaction, error) {
+	client, err := c.client.WriteTransaction(c.ctx)
+	if err != nil {
+		return nil, err
+	}
+	return &WriteTransaction{client: client, storeID: storeID, modelName: modelName}, nil
+}
+
+// Listen provides an update whenever the specified model is updated
+func (c *Client) Listen(storeID, modelName, entityID string) (chan []byte, func(), error) {
```

Comment:
Would be nice to have a typed channel to return here, but not sure if that is possible without generics.

Reply:
We can ask the client to provide a _default instance type_ of this model, and with some reflect magic unmarshal objects of that type. But that would be a `chan interface{}` anyways.

Reply:
Yea, not a ton of help I guess if the cannel is still `interface{}`. In that case, the user could just cast to their go type. The way it is now, the unmarshal into their go type. Not much different.